FHEM/10_SD_Rojaflex.pm

```diff
@@ -0,0 +1,630 @@
+		my $timelongest = $timeToOpen;
+		if ($timeToClose > $timeToOpen) {$timelongest = $timeToClose};
+		Log3 $name, 4, "$ioname: SD_Rojaflex set $name clearFav running time $timelongest s";
+		InternalTimer( (gettimeofday() + $timelongest), \&SD_Rojaflex_clearfav, $name );
+		$hash->{clearfavcount} = 0;
+	}
+	# Calculate target position and motor state
+	if ($cmd eq 'down') {
+		if ($na == 1) {$tpos = '100'}; # nicht bei "set pct xx"
+		if ($cpos ne $tpos) {$motor = 'down'}; # Wenn nicht schon unten.
+		if ($cpos eq $tpos) {$motor = 'stop'}; # Wenn unten.
+	}
+	if ($cmd eq 'up') {
+		if ($na == 1) {$tpos = '0'}; # nicht bei "set pct xx"
+		if ($cpos ne $tpos) {$motor = 'up'}; # Wenn nicht schon oben.
+		if ($cpos eq $tpos) {$motor = 'stop'}; # Wenn oben.
+	}
+	if ($cmd eq 'stop' || $cmd eq 'savefav') {
+		$motor = 'stop';
+	}
+
+	$state = $cmd;
+	Log3 $name, 3, "$ioname: SD_Rojaflex set $name $state";
+
+	readingsBeginUpdate($hash);
+	readingsBulkUpdate($hash, 'state', $state, 1);
+	if ($state ne 'clearfav' && $state ne 'gotofav') {
+		# Wenn keine PositionUpdates vom Motor kommen, setze gleich die finale Position
+		if (AttrVal($name,'bidirectional',1) eq '0') {
+			# Jump direct to the final position, because we have no position updates and set motor stop
+			$cpos = $tpos;
+			$motor = 'stop';
+			# Save current position
+			if (AttrVal($name,'inversePosition',0) eq '1') {$cpos = 100 - $cpos}; # inverse position
+			readingsBulkUpdate($hash, 'pct', $cpos, 1);
+			readingsBulkUpdate($hash, 'cpos', $cpos, 1);
+		}
+		readingsBulkUpdate($hash, 'motor', $motor, 1);
+		if (AttrVal($name,'inversePosition',0) eq '1') {$tpos = 100 - $tpos}; # inverse position
+		readingsBulkUpdate($hash, 'tpos', $tpos, 1);
+	}
+	readingsEndUpdate($hash, 1);
+
+	# channel 0 set all devices, we must update all other devices with the same housecode
+	if ($channel eq '0') {
+		foreach my $d (devspec2array("TYPE=SD_Rojaflex:FILTER=DEF=$housecode.*:FILTER=DEF!=.*_0")) {
+			Log3 $name, 3, "$ioname: SD_Rojaflex update $d $state";
+			readingsBeginUpdate($defs{$d});
+			readingsBulkUpdate($defs{$d}, 'state' , $state , 1);
+			if ($state ne 'clearfav' && $state ne 'gotofav') {
+				readingsBulkUpdate($defs{$d}, 'motor', $motor, 1);
+				readingsBulkUpdate($defs{$d}, 'tpos', $tpos, 1);
+				if (AttrVal($defs{$d}{NAME},'bidirectional',1) eq '0') {
+					readingsBulkUpdate($defs{$d}, 'pct', $cpos, 1);
+					readingsBulkUpdate($defs{$d}, 'cpos', $cpos, 1);
+				}
+			}
+			readingsEndUpdate($defs{$d}, 1);
+		}
+	}
+	return;
+}
+
+sub SD_Rojaflex_pctStop {
+	my ($name) = @_;
+	my $hash = $defs{$name};
+	RemoveInternalTimer($name);
+	CommandSet($hash, "$name stop");
+	return;
+}
+
+sub SD_Rojaflex_clearfav {
+	my ($name) = @_;
+	my $hash = $defs{$name};
+	RemoveInternalTimer($name);
+	$hash->{clearfavcount} += 1;
+	if ($hash->{clearfavcount} < 4) { # 3 mal stop senden
+		CommandSet($hash, "$name stop");
+		InternalTimer( (gettimeofday() + 1), \&SD_Rojaflex_clearfav, $name );
```

Comment:
Hier geht mir durch den Kopf einfach vier mal das Stop Kommando zu senden und das Senden die queue erledigen zu lassen.
Aber vermutlich hat es einen Grund dafür einen 1 Sekunden Timer zu setzen. Richtig?
Wenn ja, dann könnte man hier auch einfach 4x InternalTimer aufrufen?

``` 
$ttime=gettimeofday();
for (0..3) {
    InternalTimer( (++$ttime), \&SD_Rojaflex_clearfav, $name );
}
$hash->{clearfavcount} = 4;
```

naja, das RemoveInternalTimer wäre da natürlich nicht so hilfreich und müsste umgebaut werden, mir kam es nur seltsam vor, dass immer vier Timer gestellt werden, aber immer nur einer gestellt wird wenn der alte fertig ist.
An der Stelle möchte ich noch hinweisen, dass das Entfernen aller Timer die $name als Argument erhalten haben nicht gänzlich ungefährlich ist, da auch andere (Module, Nutzer) einen Timer mit diesem Argument einstellen könnten.
Vielleicht nicht wahrscheinlich, aber grundsätzlich möglich.

Reply:
Das Kommando Stop ohne definierte Pause dazwischen wurde von den Empfängern nicht akzeptiert.

$name sollte doch aber eindeutig sein - oder ist es möglich, verschiedene Devices mit gleichem Namen anzulegen?

Reply:
$name ist eindeutig, aber jeder könnte einen Timer mit $name anlegen.

Entweder in einer eigenen sub oder durch ein Modul welches Timer verwendet. Keine Ahnung ob das passiert.
Ich wollte es nur grundsätzlich erwähnen dass der Modulautor kein Exclusivrecht auf timer mit $name hat